modules/javafx.graphics/src/main/java/com/sun/javafx/application/HostServicesDelegate.java

```diff
@@ -135,22 +133,44 @@ public void showDocument(final String uri) {
                     Runtime.getRuntime().exec(
                             "rundll32 url.dll,FileProtocolHandler " + uri);
                 } else { //assume Unix or Linux
-                    String browser = null;
-                    for (String b : browsers) {
-                        if (browser == null && Runtime.getRuntime().exec(
-                                new String[]{"which", b}).getInputStream().read() != -1) {
-                            Runtime.getRuntime().exec(new String[]{browser = b, uri});
-                        }
-                    }
-                    if (browser == null) {
-                        throw new Exception("No web browser found");
-                    }
+                    showDocumentUnix(uri);
                 }
             } catch (Exception e) {
                 // should not happen
                 // dump stack for debug purpose
                 e.printStackTrace();
             }
         }
+
+        static final String[] browsers = {"google-chrome", "firefox", "opera", "konqueror", "mozilla"};
+
+        /**
+         * Try opening the document  at the given uri with the default application for its type
+         * by delegating the request to Freedesktop's {@code xdg-open} utility.
+         * If it fails, try opening using the binary executable name of various popular browsers.
+         *
+         * @param uri The URI of the document to be opened
+         *
+         * @throws Exception in case the document URI was invalid or could not be opened properly
+         */
+        private void showDocumentUnix(final String uri) throws Exception {
+            try {
+                new ProcessBuilder("xdg-open", uri).start();
```

Comment:
Now that I see what this is doing, wouldn't a simpler solution be to just take the original code and add `"xdg-open"` to the `browsers` list as the first item? You might still need a try/catch around the execution, so you could try the next item in the list if you got an exception.

In addition to being less code change, it would unify the approach between the case of using xdg-open and the specific browsers in the list.

Reply:
Indeed that would be a cleaner way to go about it! Will change to it.

Reply:
Changed it. Just slightly reformatted the array declaration as I feel it looks much nicer this way. Let me know if that's fine.

Reply:
I'll take a look when I get a chance.